user_modules/api/APITestDataGenerator.lua

```diff
@@ -0,0 +1,308 @@
+----------------------------------------------------------------------------------------------------
+-- API Test Data Generator module
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]-------------------------------------------------------------------
+local ah = require("user_modules/api/APIHelper")
+
+--[[ Module ]]--------------------------------------------------------------------------------------
+local m = {}
+
+--[[ Constants ]]-----------------------------------------------------------------------------------
+m.valueType = {
+  VALID_RANDOM = 1,
+  LOWER_IN_BOUND = 2,
+  UPPER_IN_BOUND = 3,
+  LOWER_OUT_OF_BOUND = 4,
+  UPPER_OUT_OF_BOUND = 5,
+  INVALID_TYPE = 6
+}
+
+--[[ Value generators ]]----------------------------------------------------------------------------
+math.randomseed(tonumber(tostring(os.time()):reverse():sub(1,6)))
+
+--[[ @getStringValue: Return value for 'string' data type
+--! @parameters:
+--! pTypeData: table with data type restrictions
+--! @return: value for the parameter
+--]]
+local function getStringValue(pTypeData)
+  local length
+  if pTypeData.valueType == m.valueType.LOWER_IN_BOUND then
+    length = pTypeData.minlength
+    if not length or length == 0 then length = ah.dataType.STRING.min end
```

Comment:
Is there a reason that `length == 0` is not allowed here? Empty strings are allowed by the API

Reply:
Even though API allows to have 0-length string SDL will cut off such parameter from request/response.
E.g. in case of `cloudAppVehicleID` which is String type if HMI provides `cloudAppVehicleID=""` SDL will cut off this parameter from response. And then if there are no other VD parameters this response will be treated as invalid.

```
HMI->SDL [08-01-2021 00:38:31,782]
{
    "id": 137,
    "jsonrpc": "2.0",
    "result": {
        "cloudAppVehicleID": "",
        "code": 0,
        "method": "VehicleInfo.GetVehicleData"
    }
}
 
SDL->MOB [08-01-2021 00:38:31,814] 
[rpcFunction: GetVehicleData, sessionId: 1, version: 2, frameType: 1, encryption: false, serviceType: 7, frameInfo: 0, messageId: 115, binaryDataSize: 0] : 
{
    "info": "Invalid message received from vehicle",
    "resultCode": "GENERIC_ERROR",
    "success": false
}


```

Reply:
Same is applicable for empty arrays, e.g. `{}`

Reply:
`cloudAppVehicleID` has a minlength of 1 (the property defaults to 1 if omitted), so it make sense that Core would mark it as invalid. 

It doesn't look like there is any vehicle data value which allows an empty string as of now, but that doesn't mean that it couldn't happen in the future.

Reply:
Interesting ...
I made an experimental SDL build from https://github.com/dboltovskyi/sdl_core/tree/tmp/vd_min_len with `minlength="0"` defined for `cloudAppVehicleID` and got the same result. I.e. empty-string is treated by SDL as invalid value.

Reply:
Interesting ...
I made an experimental SDL build from https://github.com/dboltovskyi/sdl_core/tree/tmp/vd_min_len branch.
It has `minlength="0"` defined for `cloudAppVehicleID` for `GetVD` and `OnVD` RPCs.
Result is different comparing to develop.
Now empty string is treated by SDL as valid value.
Though empty arrays are still invalid.
Please find update in 08a37fe  which returns back exclusion for String parameters with minlength="0"

Reply:
I think that this line would still need to be changed as well
```
if not length then length = api.dataType.STRING.min end
```
To account for the fact that `STRING.min=1`
```
  STRING = { type = "String", min = 1, max = 100 },
```

Reply:
Also, what do you mean the same happens with arrays? The same default applies, `minsize="1"` if omitted. If you set `minsize="0"`, then it will accept an empty array (although as demonstrated several places in this PR, you need to specify an empty array using `json.EMPTY_ARRAY`, since lua cannot differentiate an empty array from an empty map).

Reply:
1\) As for the `String` parameters:

There are default restrictions in `APIHelper.dataType` table.
For example if some `String` parameter have no restrictions defined (min or max) then default restrictions would take place.
Since default min for String is `1` it means the following:
   - length for `LOWER_IN_BOUND` would be be `1` (api.dataType.STRING.min)
   - length for `LOWER_OUT_OF_BOUND` would be `0` (api.dataType.STRING.min-1)

Another scenario if some `String` parameter have `minLength=0` then:
 - length for `LOWER_IN_BOUND` would be `0`
 - there would be no `LOWER_OUT_OF_BOUND` case generated

2\) As for the `Array` parameters:

Good catch for `json.EMPTY_ARRAY` const.

Please find update for points 1) and 2) in 33ecefb